### ISSUE TITLE
Misleading docs and overly complex tilesource deduction.

### ISSUE THREAD

**Issue by Aiosa** [openseadragon/openseadragon#2474]
I am thinking about refactoring the tile source creation, since it is so intricate even the docs are wrong, see https://github.com/openseadragon/openseadragon/blob/59645e3a0dbc321be0f503df4e6fb4b156c3fbba/src/tilesource.js#L58

I try to sum up the current behavior:
There are three scenarios of opening a tile source:
1) if it is a string parseable as XML or JSON, the string is converted to an object, continue with 3)
2) if it is a string, then
    - internally, a `TileSource` instance is created and it is ensured its argument `url` is set so that `getImageInfo()` gets called, which creates new tile source search based on retrieved data from the url string and calls again `configure` 
3) else if it is an inline object:
       a) if it defines `getTileUrl` property, it is considered a TileSource inline implementation and passed to new  `TileSource` instance
       where it again decides if `url` property is specified it goes to 2) else it expects other properties set (width, height...) explicitly
       b) else we rely on `determineType` that calls `supports` with null url, and data=inline object on the first compatible 
implementation (`supports(..)=true`) and we continue with `configure` call.

Now:
 - docs describe that `getImageInfo` can be overriden, which is true only in 3b) case, otherwise the parent method is always used
 - `TileSource.url` looks like value we would like to set manually, in reality this will create infinite loop if
    - you set the url value manually and / or set it as a member to the output object of `configure` and then call super constructor - which is done in most cases and becomes compulsory with `class` sugar
    - if you pass inline object with `getTileUrl`  that also defines `url` prop you immediatelly exit with `successCallback( customTileSource );` whereas 2) would wait for success callback from `TileSource` base class - but both would fire the same process... okay it ends up as no-op but a different tile source is created and finally thrown away
 - it is not clear from docs how `supports()` behaves, it has either data+url passed in case `getImageInfo` parent method gets executed, or [inline object]+null if inline object without `getTileUrl` is provided which was not clear to me at all, looks like another hack to allow plain tile image source configuration based on `type` property, which should be either standardized somehow, or done differently, I just don't like the option of passing a generic object without `getTileUrl` implementation that somehow relies on tilesource private level information/logic 
 - missing support for prioritties, e.g. supports() could return priority number so that multiple sources can compete who opens certain image, this can be useful for example if you would use custom `getImageInfo` that communicates with some infrastructure that provide multiple image servers with the same protocol but different configuration (e.g. authenticated VS public access, system-managed data records VS file heap etc..)

It took me one hour to figure this out, and I thought I knew the pipeline quite well. My motivation was the last point and the case I was looking into more advanced custom tile source integration.

**Comment by Aiosa** [comment]
I leave it here as a docs for later, RN I have no cappacity to fix this.

**Comment by iangilman** [comment]
Yes, this has become a total mess! Thank you for writing up the issue. 

I figure (when we have the bandwidth to look at this) we should define a nice clean extensible system and move to that by default, but somehow provide a deprecated version of the old system for a while for backwards compatibility.

**Comment by Aiosa** [comment]
One more issue to add (so I won't forget):

The parent implementation of `TileSource` uses `open-failed` event to notify when error happens. The handler is registered only with the case 1. - tileSource is a string.
````
            tileSource.addHandler( 'open-failed', function( event ) {
                failCallback( event );
            } );

````
However, if I write custom `TileSource` implementation, pass an _object_ that the implementation understands, then the handler is not registered and there is no mechanics to say 'I've failed' except for timeout.
````
    supports( data, url ){
        return data.type && data.type === "my-custom-source-id" && .... etc;
    }
    
    ...
    viewer.addTiledImage({
          tileSource: {
               type: "my-custom-source-id",
               ....
          },
    });
    
````

**Comment by iangilman** [comment]
Thank you for documenting these weirdnesses!

**Comment by Aiosa** [comment]
I might add here that some parts of this weird behavior were patched in the overhaul PR and documentation is under the way. It is still not optimal, but better, and more importantly, it supporst all the old tile sources still.
